src/os/hostname.zig

```diff
@@ -1,11 +1,106 @@
 const std = @import("std");
+const builtin = @import("builtin");
 const posix = std.posix;
 
 pub const HostnameParsingError = error{
     NoHostnameInUri,
     NoSpaceLeft,
 };
 
+pub const UrlParsingError = error{
+    HostnameIsNotMacAddress,
+    InvalidFormat,
+    InvalidPort,
+    NoSchemeProvided,
+    UnexpectedCharacter,
+};
+
+fn isUriPathSeparator(c: u8) bool {
+    return switch (c) {
+        '?', '#' => true,
+        else => false,
+    };
+}
+
+fn isValidMacAddress(mac_address: []const u8) bool {
+    // A valid mac address has 6 two-character components with 5 colons, e.g. 12:34:56:ab:cd:ef.
+    if (mac_address.len != 17) {
+        return false;
+    }
+
+    for (mac_address, 0..) |c, i| {
+        if ((i + 1) % 3 == 0) {
+            if (c != ':') {
+                return false;
+            }
+        } else {
+            if (!std.mem.containsAtLeastScalar(u8, "0123456789ABCDEFabcdef", 1, c)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+/// Parses the provided url to a `std.Uri` struct. This is very specific to getting hostname and
+/// path information for Ghostty's PWD reporting functionality. Takes into account that on macOS
+/// the url passed to this function might have a mac address as its hostname and parses it
+/// correctly.
+pub fn parseUrl(url: []const u8) UrlParsingError!std.Uri {
+    return std.Uri.parse(url) catch |e| {
+        // The mac-address-as-hostname issue is specific to macOS so we just return an error if we
+        // hit it on other platforms.
+        comptime if (builtin.os.tag != .macos) return e;
+
+        // It's possible this is a mac address on macOS where the last 2 characters in the
+        // address are non-digits, e.g. 'ff', and thus an invalid port.
+        //
+        // Example: file://12:34:56:78:90:12/path/to/file
+        if (e != error.InvalidPort) return e;
+
+        const url_without_scheme_start = std.mem.indexOf(u8, url, "://") orelse {
+            return error.NoSchemeProvided;
+        };
+        const scheme = url[0..url_without_scheme_start];
+        const url_without_scheme = url[url_without_scheme_start + 3 ..];
+
+        // The first '/' after the scheme marks the end of the hostname. If the first '/'
+        // following the end of the scheme is not at the right position this is not a
+        // valid mac address.
+        const mac_address_length = 17;
```

Comment:
This could be moved to file scope and used within `isValidMacAddress`. I suppose we could also move the MAC-specific bits to their own `mac.zig`, but I don't know if that's particularly useful yet given this is the only place we use those routines.

Reply:
oh, good point! Moved in 6ed94b00346a402d85ac83a9585de17ced93f351 🎉 

> I suppose we could also move the MAC-specific bits to their own `mac.zig`, but I don't know if that's particularly useful yet given this is the only place we use those routines.

Yeah I deliberately didn't export those functions since they're specific to this particular use in just this file. I'm happy to make them a module but I don't think there's a benefit to it at this point 🤷‍♂️